File: src/main/updates/update-service.ts
import { autoUpdater as defaultAutoUpdater } from 'electron-updater';
import type { UpdateInfo as ElectronUpdateInfo, UpdateCheckResult } from 'electron-updater';
import type { UpdateDownloadProgress, UpdateInfo, UpdateInstallResult, UpdateState } from '../../shared/update';
import { createInitialUpdateState } from '../../shared/update';
import { appendUpdateInstallDiagnostic, type UpdateInstallDiagnosticEvent } from './update-install-diagnostics';
import { launchWindowsUpdateInstaller, type UpdateInstallerLaunchReason } from './update-installer-launcher';

type ElectronDownloadProgress = {
  transferred?: number;
  total?: number;
  percent?: number;
};

export type ElectronUpdaterAdapter = {
  autoDownload: boolean;
  autoInstallOnAppQuit: boolean;
  checkForUpdates(): Promise<UpdateCheckResult | null>;
  downloadUpdate(): Promise<Array<string>>;
  on(event: string, listener: (...args: unknown[]) => void): void;
};

export type UpdateServiceOptions = {
  currentVersion: string;
  isPackaged: boolean;
  platform: NodeJS.Platform;
  autoUpdater?: ElectronUpdaterAdapter;
  launchInstaller?: typeof launchWindowsUpdateInstaller;
  now?: () => Date;
  diagnosticsFilePath?: string | null;
  setInterval?: typeof setInterval;
  clearInterval?: typeof clearInterval;
  setTimeout?: typeof setTimeout;
  clearTimeout?: typeof clearTimeout;
  onStateChanged?: (state: UpdateState) => void;
};

type UpdateOperation = 'idle' | 'checking' | 'downloading';

export const UPDATE_POLL_INTERVAL_MS = 60 * 60 * 1000;
export const INITIAL_UPDATE_POLL_DELAY_MS = 30 * 1000;

export class UpdateService {
  private readonly isPackaged: boolean;
  private readonly platform: NodeJS.Platform;
  private readonly autoUpdater: ElectronUpdaterAdapter;
  private readonly launchInstaller: typeof launchWindowsUpdateInstaller;
  private readonly now: () => Date;
  private readonly diagnosticsFilePath: string | null;
  private readonly setPollInterval: typeof setInterval;
  private readonly clearPollInterval: typeof clearInterval;
  private readonly setPollTimeout: typeof setTimeout;
  private readonly clearPollTimeout: typeof clearTimeout;
  private readonly notifyStateChanged: (state: UpdateState) => void;
  private state: UpdateState;
  private operation: UpdateOperation = 'idle';
  private checkingPromise: Promise<UpdateState> | null = null;
  private downloadPromise: Promise<UpdateState> | null = null;
  private downloadedInstallerPath: string | null = null;
  private pollInterval: NodeJS.Timeout | null = null;
  private initialPollTimeout: NodeJS.Timeout | null = null;

  constructor(options: UpdateServiceOptions) {
    this.isPackaged = options.isPackaged;
    this.platform = options.platform;
    this.autoUpdater = options.autoUpdater ?? defaultAutoUpdater;
    this.launchInstaller = options.launchInstaller ?? launchWindowsUpdateInstaller;
    this.now = options.now ?? (() => new Date());
    this.diagnosticsFilePath = options.diagnosticsFilePath ?? null;
    this.setPollInterval = options.setInterval ?? setInterval;
    this.clearPollInterval = options.clearInterval ?? clearInterval;
    this.setPollTimeout = options.setTimeout ?? setTimeout;
    this.clearPollTimeout = options.clearTimeout ?? clearTimeout;
    this.notifyStateChanged = options.onStateChanged ?? (() => undefined);
    this.state = createInitialUpdateState(options.currentVersion);

    this.autoUpdater.autoDownload = false;
    this.autoUpdater.autoInstallOnAppQuit = false;
    this.registerUpdaterEvents();
  }

  getState(): UpdateState {
    return cloneState(this.state);
  }

  startAutomaticUpdateChecks(): void {
    if (this.unsupportedReason()) return;
    if (this.pollInterval || this.initialPollTimeout) return;

    this.initialPollTimeout = this.setPollTimeout(() => {
      this.initialPollTimeout = null;
      void this.runAutomaticUpdateCheck();
    }, INITIAL_UPDATE_POLL_DELAY_MS);

    this.pollInterval = this.setPollInterval(() => {
      void this.runAutomaticUpdateCheck();
    }, UPDATE_POLL_INTERVAL_MS);
  }

  stopAutomaticUpdateChecks(): void {
    if (this.initialPollTimeout) {
      this.clearPollTimeout(this.initialPollTimeout);
      this.initialPollTimeout = null;
    }

    if (this.pollInterval) {
      this.clearPollInterval(this.pollInterval);
      this.pollInterval = null;
    }
  }

  checkForUpdates(): Promise<UpdateState> {
    const unsupportedReason = this.unsupportedReason();
    if (unsupportedReason) {
      this.setState({
        ...this.state,
        info: {
          ...this.state.info,
          checkedAt: this.now().toISOString(),
          status: 'check_failed',
          reason: unsupportedReason
        },
        download: createIdleDownload()
      });
      return Promise.resolve(this.getState());
    }

    if (this.checkingPromise) return Promise.resolve(this.getState());

    this.operation = 'checking';
    this.downloadedInstallerPath = null;
    this.setState({
      info: {
        ...this.state.info,
        latestVersion: null,
        releaseName: null,
        releaseNotes: null,
        checkedAt: null,
        status: 'checking',
        reason: undefined
      },
      download: createIdleDownload()
    });

    this.checkingPromise = this.autoUpdater
      .checkForUpdates()
      .then(() => this.getState())
      .catch((error) => {
        console.error('Switchify update check failed.', error);
        this.setState({
          ...this.state,
          info: {
            ...this.state.info,
            checkedAt: this.now().toISOString(),
            status: 'check_failed',
            reason: 'network_error'
          }
        });
        return this.getState();
      })
      .finally(() => {
        this.operation = 'idle';
        this.checkingPromise = null;
      });

    return this.checkingPromise;
  }

  async downloadUpdate(): Promise<UpdateState> {
    const unsupportedReason = this.unsupportedReason();
    if (unsupportedReason) {
      this.setState({
        ...this.state,
        download: {
          ...createIdleDownload(),
          status: 'download_failed',
          reason: unsupportedReason
        }
      });
      return this.getState();
    }

    if (this.downloadPromise) return this.getState();

    if (this.state.info.status !== 'update_available') {
      this.setState({
        ...this.state,
        download: {
          ...createIdleDownload(),
          status: 'download_failed',
          reason: 'not_available'
        }
      });
      return this.getState();
    }

    this.operation = 'downloading';
    this.downloadedInstallerPath = null;
    this.setState({
      ...this.state,
      download: {
        status: 'downloading',
        downloadedBytes: 0,
        totalBytes: null,
        percent: null
      }
    });

    this.downloadPromise = this.autoUpdater
      .downloadUpdate()
      .then((downloadedPaths) => {
        this.downloadedInstallerPath = firstInstallerPath(downloadedPaths);
        return this.getState();
      })
      .catch((error) => {
        console.error('Switchify update download failed.', error);
        this.downloadedInstallerPath = null;
        this.setState({
          ...this.state,
          download: {
            ...this.state.download,
            status: 'download_failed',
            reason: 'network_error'
          }
        });
        return this.getState();
      })
      .finally(() => {
        this.operation = 'idle';
        this.downloadPromise = null;
      });

    return this.downloadPromise;
  }

  async installDownloadedUpdate(): Promise<UpdateInstallResult> {
    this.recordInstallDiagnostic('install_requested');
    const unsupportedReason = this.unsupportedReason();
    if (unsupportedReason) {
      return { ok: false, reason: unsupportedReason };
    }

    if (this.state.download.status !== 'downloaded') {
      return { ok: false, reason: 'not_downloaded' };
    }

    const result = await this.launchInstaller({ installerPath: this.downloadedInstallerPath });

    if (!result.ok) {
      console.error('Switchify update installer could not be opened.', result.reason);
      this.recordInstallDiagnostic(diagnosticEventForLaunchFailure(result.reason), result.reason);
      return { ok: false, reason: result.reason };
    }

    this.recordInstallDiagnostic('installer_started');
    return { ok: true };
  }

  recordInstallCancelled(): void {
    this.recordInstallDiagnostic('confirmation_cancelled');
  }

  private registerUpdaterEvents(): void {
    this.autoUpdater.on('update-available', (rawInfo) => {
      const info = rawInfo as ElectronUpdateInfo;
      this.downloadedInstallerPath = null;
      this.setState({
        info: updateInfo(this.state.info, info, {
          checkedAt: this.now().toISOString(),
          status: 'update_available',
          reason: undefined
        }),
        download: createIdleDownload()
      });
    });

    this.autoUpdater.on('update-not-available', (rawInfo) => {
      const info = rawInfo as ElectronUpdateInfo;
      this.downloadedInstallerPath = null;
      this.setState({
        ...this.state,
        info: updateInfo(this.state.info, info, {
          checkedAt: this.now().toISOString(),
          latestVersion: info.version || this.state.info.currentVersion,
          status: 'up_to_date',
          reason: undefined
        }),
        download: createIdleDownload()
      });
    });

    this.autoUpdater.on('download-progress', (rawProgress) => {
      const progress = rawProgress as ElectronDownloadProgress;
      const downloadedBytes = Number.isFinite(progress.transferred) ? Number(progress.transferred) : 0;
      const totalBytes = Number.isFinite(progress.total) && Number(progress.total) > 0 ? Number(progress.total) : null;
      const percent =
        Number.isFinite(progress.percent) && Number(progress.percent) >= 0
          ? Math.min(100, Math.round(Number(progress.percent)))
          : totalBytes
            ? Math.min(100, Math.round((downloadedBytes / totalBytes) * 100))
            : null;

      this.setState({
        ...this.state,
        download: {
          status: 'downloading',
          downloadedBytes,
          totalBytes,
          percent
        }
      });
    });

    this.autoUpdater.on('update-downloaded', (rawInfo) => {
      const info = rawInfo as ElectronUpdateInfo;
      this.setState({
        ...this.state,
        info: updateInfo(this.state.info, info),
        download: {
          status: 'downloaded',
          downloadedBytes: this.state.download.downloadedBytes,
          totalBytes: this.state.download.totalBytes,
          percent: 100
        }
      });
    });

    this.autoUpdater.on('error', (error) => {
      console.error('Switchify updater error.', error);
      if (this.operation === 'downloading') {
        this.downloadedInstallerPath = null;
        this.setState({
          ...this.state,
          download: {
            ...this.state.download,
            status: 'download_failed',
            reason: 'network_error'
          }
        });
        return;
      }

      this.setState({
        ...this.state,
        info: {
          ...this.state.info,
          checkedAt: this.now().toISOString(),
          status: 'check_failed',
          reason: 'network_error'
        }
      });
    });
  }

  private async runAutomaticUpdateCheck(): Promise<void> {
    if (this.operation !== 'idle') return;
    if (this.state.download.status === 'downloading' || this.state.download.status === 'downloaded') return;

    await this.checkForUpdates();
  }

  private setState(state: UpdateState): void {
    this.state = state;
    this.notifyStateChanged(this.getState());
  }

  private unsupportedReason(): 'not_packaged' | 'not_supported' | null {
    if (!this.isPackaged) return 'not_packaged';
    if (this.platform !== 'win32') return 'not_supported';
    return null;
  }

  private recordInstallDiagnostic(event: UpdateInstallDiagnosticEvent, reason?: string): void {
    if (!this.diagnosticsFilePath) return;

    appendUpdateInstallDiagnostic(this.diagnosticsFilePath, {
      event,
      at: this.now().toISOString(),
      version: this.state.info.currentVersion,
      ...(reason ? { reason } : {})
    });
  }

}

function firstInstallerPath(paths: string[]): string | null {
  return paths.find((path) => path.toLowerCase().endsWith('.exe')) ?? paths[0] ?? null;
}

function diagnosticEventForLaunchFailure(reason: UpdateInstallerLaunchReason): UpdateInstallDiagnosticEvent {
  switch (reason) {
    case 'installer_unavailable':
      return 'installer_missing';
    case 'installer_launch_failed':
      return 'installer_launch_failed';
  }
}


function updateInfo(
  previous: UpdateInfo,
  info: ElectronUpdateInfo,
  patch: Partial<UpdateInfo> = {}
): UpdateInfo {
  return {
    ...previous,
    latestVersion: info.version || previous.latestVersion,
    releaseName: toNullableString(info.releaseName),
    releaseNotes: normalizeReleaseNotes(info.releaseNotes),
    ...patch
  };
}

function normalizeReleaseNotes(notes: unknown): string | null {
  if (typeof notes === 'string') return notes;
  if (Array.isArray(notes)) {
    const text = notes
      .map((note) => {
        if (typeof note === 'string') return note;
        if (note && typeof note === 'object' && 'note' in note && typeof note.note === 'string') return note.note;
        return null;
      })
      .filter(Boolean)
      .join('\n\n');
    return text.length > 0 ? text : null;
  }
  return null;
}

function toNullableString(value: unknown): string | null {
  return typeof value === 'string' ? value : null;
}

function createIdleDownload(): UpdateDownloadProgress {
  return {
    status: 'idle',
    downloadedBytes: 0,
    totalBytes: null,
    percent: null
  };
}

function cloneState(state: UpdateState): UpdateState {
  return {
    info: { ...state.info },
    download: { ...state.download }
  };
}

File: src-dotnet/SwitchifyPc.Core/Updates/UpdateService.cs
namespace SwitchifyPc.Core.Updates;

public interface IUpdateBackend
{
    Task<UpdateCheckOutcome> CheckForUpdatesAsync(string currentVersion, CancellationToken cancellationToken = default);

    Task<UpdateDownloadOutcome> DownloadUpdateAsync(
        AvailableUpdate update,
        IProgress<UpdateDownloadSnapshot> progress,
        CancellationToken cancellationToken = default);
}

public interface IUpdateInstallerLauncher
{
    Task<UpdateInstallerLaunchResult> LaunchAsync(string? installerPath, CancellationToken cancellationToken = default);
}

public interface IUpdatePollScheduler
{
    IDisposable ScheduleOnce(TimeSpan delay, Func<Task> callback);

    IDisposable ScheduleRecurring(TimeSpan interval, Func<Task> callback);
}

public sealed class TimerUpdatePollScheduler : IUpdatePollScheduler
{
    public IDisposable ScheduleOnce(TimeSpan delay, Func<Task> callback)
    {
        Timer? timer = null;
        timer = new Timer(_ =>
        {
            timer?.Dispose();
            _ = callback();
        }, null, delay, Timeout.InfiniteTimeSpan);

        return timer;
    }

    public IDisposable ScheduleRecurring(TimeSpan interval, Func<Task> callback)
    {
        return new Timer(_ => _ = callback(), null, interval, interval);
    }
}

public sealed class UpdateService
{
    public static readonly TimeSpan PollInterval = TimeSpan.FromHours(1);
    public static readonly TimeSpan InitialPollDelay = TimeSpan.FromSeconds(30);

    private readonly bool isPackaged;
    private readonly string platform;
    private readonly IUpdateBackend backend;
    private readonly IUpdateInstallerLauncher installerLauncher;
    private readonly IUpdatePollScheduler scheduler;
    private readonly Func<DateTimeOffset> now;
    private readonly Action<UpdateState> onStateChanged;
    private readonly object gate = new();

    private UpdateState state;
    private string? downloadedInstallerPath;
    private AvailableUpdate? availableUpdate;
    private Task<UpdateState>? checkingTask;
    private Task<UpdateState>? downloadTask;
    private UpdateOperation operation = UpdateOperation.Idle;
    private IDisposable? initialPoll;
    private IDisposable? recurringPoll;

    public UpdateService(UpdateServiceOptions options)
    {
        ArgumentNullException.ThrowIfNull(options);
        isPackaged = options.IsPackaged;
        platform = options.Platform;
        backend = options.Backend;
        installerLauncher = options.InstallerLauncher;
        scheduler = options.Scheduler ?? new TimerUpdatePollScheduler();
        now = options.Now ?? (() => DateTimeOffset.UtcNow);
        onStateChanged = options.OnStateChanged ?? (_ => { });
        state = UpdateState.CreateInitial(options.CurrentVersion);
    }

    public UpdateState GetState()
    {
        lock (gate)
        {
            return state;
        }
    }

    public void StartAutomaticUpdateChecks()
    {
        if (UnsupportedReason() is not null) return;

        lock (gate)
        {
            if (initialPoll is not null || recurringPoll is not null) return;
            initialPoll = scheduler.ScheduleOnce(InitialPollDelay, RunAutomaticUpdateCheckAsync);
            recurringPoll = scheduler.ScheduleRecurring(PollInterval, RunAutomaticUpdateCheckAsync);
        }
    }

    public void StopAutomaticUpdateChecks()
    {
        IDisposable? oneShot;
        IDisposable? recurring;
        lock (gate)
        {
            oneShot = initialPoll;
            recurring = recurringPoll;
            initialPoll = null;
            recurringPoll = null;
        }

        oneShot?.Dispose();
        recurring?.Dispose();
    }

    public Task<UpdateState> CheckForUpdatesAsync(CancellationToken cancellationToken = default)
    {
        UpdateFailureReason? unsupported = UnsupportedReason();
        if (unsupported is not null)
        {
            SetState(GetState() with
            {
                Info = GetState().Info with
                {
                    CheckedAt = now(),
                    Status = UpdateCheckStatus.CheckFailed,
                    Reason = unsupported
                },
                Download = UpdateState.CreateIdleDownload()
            });
            return Task.FromResult(GetState());
        }

        lock (gate)
        {
            if (checkingTask is { IsCompleted: false }) return checkingTask;
            checkingTask = null;
            operation = UpdateOperation.Checking;
            downloadedInstallerPath = null;
            availableUpdate = null;
            state = state with
            {
                Info = state.Info with
                {
                    LatestVersion = null,
                    ReleaseName = null,
                    ReleaseNotes = null,
                    CheckedAt = null,
                    Status = UpdateCheckStatus.Checking,
                    Reason = null
                },
                Download = UpdateState.CreateIdleDownload()
            };
            onStateChanged(state);
            checkingTask = RunCheckAsync(cancellationToken);
            return checkingTask;
        }
    }

    public Task<UpdateState> DownloadUpdateAsync(CancellationToken cancellationToken = default)
    {
        UpdateFailureReason? unsupported = UnsupportedReason();
        if (unsupported is not null)
        {
            SetState(GetState() with
            {
                Download = UpdateState.CreateIdleDownload() with
                {
                    Status = UpdateDownloadStatus.DownloadFailed,
                    Reason = unsupported
                }
            });
            return Task.FromResult(GetState());
        }

        lock (gate)
        {
            if (downloadTask is { IsCompleted: false }) return downloadTask;
            downloadTask = null;
            if (state.Info.Status != UpdateCheckStatus.UpdateAvailable || availableUpdate is null)
            {
                state = state with
                {
                    Download = UpdateState.CreateIdleDownload() with
                    {
                        Status = UpdateDownloadStatus.DownloadFailed,
                        Reason = UpdateFailureReason.NotAvailable
                    }
                };
                onStateChanged(state);
                return Task.FromResult(state);
            }

            operation = UpdateOperation.Downloading;
            downloadedInstallerPath = null;
            state = state with
            {
                Download = new UpdateDownloadProgress(UpdateDownloadStatus.Downloading, 0, null, null)
            };
            onStateChanged(state);
            downloadTask = RunDownloadAsync(availableUpdate, cancellationToken);
            return downloadTask;
        }
    }

    public async Task<UpdateInstallResult> InstallDownloadedUpdateAsync(CancellationToken cancellationToken = default)
    {
        UpdateFailureReason? unsupported = UnsupportedReason();
        if (unsupported == UpdateFailureReason.NotPackaged) return UpdateInstallResult.Failure(UpdateInstallFailureReason.NotPackaged);
        if (unsupported == UpdateFailureReason.NotSupported) return UpdateInstallResult.Failure(UpdateInstallFailureReason.NotSupported);

        string? installerPath;
        lock (gate)
        {
            if (state.Download.Status != UpdateDownloadStatus.Downloaded) return UpdateInstallResult.Failure(UpdateInstallFailureReason.NotDownloaded);
            installerPath = downloadedInstallerPath;
        }

        UpdateInstallerLaunchResult result = await installerLauncher.LaunchAsync(installerPath, cancellationToken).ConfigureAwait(false);
        return result.Ok
            ? UpdateInstallResult.Success()
            : UpdateInstallResult.Failure(result.Reason ?? UpdateInstallFailureReason.InstallerLaunchFailed);
    }

    private async Task<UpdateState> RunCheckAsync(CancellationToken cancellationToken)
    {
        try
        {
            UpdateCheckOutcome outcome = await backend.CheckForUpdatesAsync(GetState().Info.CurrentVersion, cancellationToken).ConfigureAwait(false);
            if (outcome.FailureReason is not null)
            {
                SetState(GetState() with
                {
                    Info = GetState().Info with
                    {
                        CheckedAt = now(),
                        Status = UpdateCheckStatus.CheckFailed,
                        Reason = outcome.FailureReason
                    }
                });
                return GetState();
            }

            if (outcome.UpdateAvailable && outcome.Update is not null)
            {
                lock (gate)
                {
                    availableUpdate = outcome.Update;
                }

                SetState(GetState() with
                {
                    Info = GetState().Info with
                    {
                        LatestVersion = outcome.Update.Version,
                        ReleaseName = outcome.Update.ReleaseName,
                        ReleaseNotes = outcome.Update.ReleaseNotes,
                        CheckedAt = now(),
                        Status = UpdateCheckStatus.UpdateAvailable,
                        Reason = null
                    },
                    Download = UpdateState.CreateIdleDownload()
                });
                return GetState();
            }

            SetState(GetState() with
            {
                Info = GetState().Info with
                {
                    LatestVersion = GetState().Info.CurrentVersion,
                    CheckedAt = now(),
                    Status = UpdateCheckStatus.UpToDate,
                    Reason = null
                },
                Download = UpdateState.CreateIdleDownload()
            });
            return GetState();
        }
        catch
        {
            SetState(GetState() with
            {
                Info = GetState().Info with
                {
                    CheckedAt = now(),
                    Status = UpdateCheckStatus.CheckFailed,
                    Reason = UpdateFailureReason.NetworkError
                }
            });
            return GetState();
        }
        finally
        {
            lock (gate)
            {
                operation = UpdateOperation.Idle;
                checkingTask = null;
            }
        }
    }

    private async Task<UpdateState> RunDownloadAsync(AvailableUpdate update, CancellationToken cancellationToken)
    {
        try
        {
            InlineProgress<UpdateDownloadSnapshot> progress = new(snapshot =>
            {
                SetState(GetState() with
                {
                    Download = new UpdateDownloadProgress(
                        UpdateDownloadStatus.Downloading,
                        snapshot.DownloadedBytes,
                        snapshot.TotalBytes,
                        snapshot.Percent is null ? null : Math.Clamp(snapshot.Percent.Value, 0, 100))
                });
            });

            UpdateDownloadOutcome outcome = await backend.DownloadUpdateAsync(update, progress, cancellationToken).ConfigureAwait(false);
            if (outcome.FailureReason is not null || string.IsNullOrWhiteSpace(outcome.InstallerPath))
            {
                lock (gate)
                {
                    downloadedInstallerPath = null;
                }

                SetState(GetState() with
                {
                    Download = GetState().Download with
                    {
                        Status = UpdateDownloadStatus.DownloadFailed,
                        Reason = outcome.FailureReason ?? UpdateFailureReason.InvalidUpdate
                    }
                });
                return GetState();
            }

            lock (gate)
            {
                downloadedInstallerPath = outcome.InstallerPath;
            }

            SetState(GetState() with
            {
                Download = GetState().Download with
                {
                    Status = UpdateDownloadStatus.Downloaded,
                    Percent = 100,
                    Reason = null
                }
            });
            return GetState();
        }
        catch
        {
            lock (gate)
            {
                downloadedInstallerPath = null;
            }

            SetState(GetState() with
            {
                Download = GetState().Download with
                {
                    Status = UpdateDownloadStatus.DownloadFailed,
                    Reason = UpdateFailureReason.NetworkError
                }
            });
            return GetState();
        }
        finally
        {
            lock (gate)
            {
                operation = UpdateOperation.Idle;
                downloadTask = null;
            }
        }
    }

    private async Task RunAutomaticUpdateCheckAsync()
    {
        lock (gate)
        {
            if (operation != UpdateOperation.Idle) return;
            if (state.Download.Status is UpdateDownloadStatus.Downloading or UpdateDownloadStatus.Downloaded) return;
        }

        await CheckForUpdatesAsync().ConfigureAwait(false);
    }

    private void SetState(UpdateState next)
    {
        lock (gate)
        {
            state = next;
        }

        onStateChanged(next);
    }

    private UpdateFailureReason? UnsupportedReason()
    {
        if (!isPackaged) return UpdateFailureReason.NotPackaged;
        if (!string.Equals(platform, "win32", StringComparison.OrdinalIgnoreCase)) return UpdateFailureReason.NotSupported;
        return null;
    }

    private enum UpdateOperation
    {
        Idle,
        Checking,
        Downloading
    }

    private sealed class InlineProgress<T>(Action<T> handler) : IProgress<T>
    {
        public void Report(T value)
        {
            handler(value);
        }
    }
}

public sealed record UpdateServiceOptions(
    string CurrentVersion,
    bool IsPackaged,
    string Platform,
    IUpdateBackend Backend,
    IUpdateInstallerLauncher InstallerLauncher,
    IUpdatePollScheduler? Scheduler = null,
    Func<DateTimeOffset>? Now = null,
    Action<UpdateState>? OnStateChanged = null);
